plugins/statistics.koplugin/main.lua

```diff
@@ -598,6 +603,38 @@ function ReaderStatistics:upgradeDBto20201022(conn)
     conn:exec("PRAGMA user_version=20201022;")
 end
 
+function ReaderStatistics:upgradeDBto20221022(conn)
+    conn:exec([[
+        -- We make the index on book's (title, author, md5) unique in order to sync dbs
+        -- First we remove duplicates (most likely none existant?)
+        UPDATE page_stat_data SET id_book = (
+            SELECT map.min_id FROM (
+                SELECT id, (
+                    SELECT min(id) FROM book b2
+                    WHERE (book.title, book.authors, book.md5) = (b2.title, b2.authors, b2.md5)
+                ) as min_id
+                FROM book WHERE book.id >= min_id
+            ) as map WHERE page_stat_data.id_book = map.id
+        );
+        DELETE FROM book WHERE id > (
+            SELECT MIN(id) FROM book b2
+            WHERE (book.title, book.authors, book.md5) = (b2.title, b2.authors, b2.md5)
+        );
+        -- Then we recompute the book statistics based on merged books
+        UPDATE book SET (total_read_pages, total_read_time) =
+        (SELECT count(DISTINCT page),
+                sum(duration)
+         FROM   page_stat
+         WHERE  id_book = book.id);
+
+        -- Finally we update the index to be unique
+        DROP INDEX IF EXISTS book_title_authors_md5;
+        CREATE UNIQUE INDEX book_title_authors_md5 ON book(title, authors, md5);]])
```

Comment:
I haven't checked, but:
- can it happen in the code that we would want to insert a duplicate of (title, authors, md5) ?
- what will happen if we do: the insertion will fail because unique key exists, what would the followup code do? Refetch the id_book and get the existing one ?

Reply:
It's already been dealt with, we can't insert a duplicate :)
https://github.com/koreader/koreader/blob/cf643c453bcbbef0d7986c625112d708558d5471/plugins/statistics.koplugin/main.lua#L619-L633
and
https://github.com/koreader/koreader/blob/cf643c453bcbbef0d7986c625112d708558d5471/plugins/statistics.koplugin/main.lua#L758-L773